Sources/Basics/Observability.swift

```diff
@@ -663,3 +705,43 @@ extension ObservabilityMetadata {
         }
     }
 }
+
+// MARK: - Progress
+
+public protocol ScopeProgressHandler {
+    func handleProgress(scope: ObservabilityScope, step: Int64, total: Int64, unit: String?, description: String?)
+}
+
+public protocol ProgressEmitterProtocol {
+    func emit(step: Int64, total: Int64, unit: String?, description: String?)
+}
+
+extension ProgressEmitterProtocol {
+    public func emit(step: Int, total: Int, unit: String?, description: String?) {
+        self.emit(step: Int64(step), total: Int64(total), unit: unit, description: description)
+    }
+}
+
+// MARK: - User output
+
+public protocol ScopeOutputHandler {
+    func handleOutput(scope: ObservabilityScope, output: String)
+}
+
+public protocol OutputEmitterProtocol {
+    func emit(output: String)
+}
+
+extension OutputEmitterProtocol {
+    public func emit(output bytes: [UInt8]) {
+        if let output = String(bytes: bytes, encoding: .utf8) {
```

Comment:
It might be better to use the decoding form that patches up invalid UTF-8, since otherwise output that contains invalid UTF-8 just disappears.

Also, should this split up by line, or is that done at the higher level?  If this is not by line, then we should probably keep as data until it's been split up, since we might break multi-byte UTF-8 sequences otherwise.